ddl/partition.go

```diff
@@ -1940,27 +1953,139 @@ func onTruncateTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (int64, e
 		return ver, errors.Trace(dbterror.ErrPartitionMgmtOnNonpartitioned)
 	}
 
-	newPartitions := make([]model.PartitionDefinition, 0, len(oldIDs))
-	for _, oldID := range oldIDs {
+	switch job.SchemaState {
+	case model.StatePublic:
+		// Step1: generate new partition ids
+		newDefinitions := make([]model.PartitionDefinition, 0, len(pi.Definitions))
 		for i := 0; i < len(pi.Definitions); i++ {
-			def := &pi.Definitions[i]
-			if def.ID == oldID {
-				pid, err1 := t.GenGlobalID()
-				if err1 != nil {
-					return ver, errors.Trace(err1)
+			newDef := pi.Definitions[i].Clone()
+			for _, oldID := range oldIDs {
+				if newDef.ID == oldID {
+					pid, err1 := t.GenGlobalID()
```

Comment:
Maybe allocate the new global ids already in the alter table client session, like [this](https://github.com/pingcap/tidb/blob/a085eff3da5c5c195985a029113d3e0c872a7192/ddl/ddl_api.go#L6269-L6276)? (Recently fixed https://github.com/pingcap/tidb/issues/42940)